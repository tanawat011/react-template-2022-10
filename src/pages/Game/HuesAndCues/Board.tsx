import type { RoomPlayer } from './_type'
import type { HuesAndCues } from 'data/HuesAndCues'

import { TwCol, TwRow } from 'components/Common'
import { getColTitles, getRowTitles, huesAndCues } from 'data/HuesAndCues'

import { CellClickable, CellTransparent, Marker } from './Common'
import { setPlayerInTheRoom } from './services'

const ColTitles = () => {
  const titles = ['', ...getColTitles(), '']

  return (
    <TwRow>
      {titles.map((title, index) => {
        return <CellTransparent key={`col-title-${title}-${index}`}>{title}</CellTransparent>
      })}
    </TwRow>
  )
}

const RowTitles = () => {
  const titles = getRowTitles()

  return (
    <TwCol>
      {titles.map((title, index) => {
        return <CellTransparent key={`row-title-${title}-${index}`}>{title}</CellTransparent>
      })}
    </TwCol>
  )
}

type Prop = {
  roomId: string
  currentPlayer: RoomPlayer
  playersInRoom: RoomPlayer[]
  updateCurrentPlayer: (currentPlayer: RoomPlayer) => Promise<void>
}

export const Board: React.FC<Prop> = ({
  roomId,
  currentPlayer,
  playersInRoom,
  updateCurrentPlayer,
}) => {
  const handlePlayerCells = async (item: HuesAndCues) => {
    const countTurn = currentPlayer?.countTurn || 0

    if (currentPlayer && !currentPlayer.hinter && currentPlayer.isMyTurn && countTurn < 1) {
      let cells = currentPlayer.cells
      const isDuplicateCell = cells.includes(item.id)

      if (isDuplicateCell) {
        cells = cells.filter((cell) => cell !== item.id)
      } else {
        cells.push(item.id)
      }

      // * In case has 2 players in the room
      if (playersInRoom.length === 2) {
        const _currentPlayer: RoomPlayer = {
          ...currentPlayer,
          cells,
          countTurn: countTurn + 1,
        }

        await updateCurrentPlayer(_currentPlayer)

        return
      }

      const playersInRoomSort = playersInRoom.sort((a, b) => a.number - b.number)

      let nextNumber = currentPlayer.number // 2
      let nextPlayerTurn = playersInRoomSort[nextNumber]

      // * In case the last player in the room
      if (!nextPlayerTurn) {
        nextNumber = 0
        nextPlayerTurn = playersInRoomSort[nextNumber]
      }

      // * In case the next player is the hinter
      if (nextPlayerTurn && nextPlayerTurn.hinter) {
        nextNumber += 1
        nextPlayerTurn = playersInRoomSort[nextNumber] || playersInRoomSort[0]
      }

      const _nextPlayerTurn: RoomPlayer = {
        ...nextPlayerTurn,
        isMyTurn: true,
      }

      await setPlayerInTheRoom(roomId, _nextPlayerTurn)

      const _currentPlayer: RoomPlayer = {
        ...currentPlayer,
        cells,
        isMyTurn: false,
        countTurn: countTurn + 1,
      }

      await updateCurrentPlayer(_currentPlayer)
    }
  }

  return (
    <>
      <ColTitles />
      <TwRow>
        <RowTitles />

        <TwCol>
          {huesAndCues.map((items, index) => {
            return (
              <TwRow key={`hues-and-cues-${index}`}>
                {items.map((cell) => {
                  const cellId = cell.id
                  const cellColor = cell.color

                  const player = playersInRoom.find((u) => u.cells.includes(cellId))

                  const hinter = playersInRoom.find((u) => u.hinter && !!u.result)
                  const isHint = hinter?.result === cellId

                  const isShowResultForHinter = currentPlayer.hinter
                  const isShowResultForPlayer = !currentPlayer.hinter && hinter?.isSubmitResult
                  const isShowResult = isHint && (isShowResultForHinter || isShowResultForPlayer)

                  return (
                    <CellClickable
                      key={`${cellColor}-${cellId}`}
                      className={cellColor}
                      onClick={() => handlePlayerCells(cell)}
                    >
                      {player?.color && <Marker className={player.color} />}
                      {isShowResult && (
                        <div className='absolute w-28 h-28 border-4 border-solid border-white' />
                      )}
                    </CellClickable>
                  )
                })}
              </TwRow>
            )
          })}
        </TwCol>

        <RowTitles />
      </TwRow>
      <ColTitles />
    </>
  )
}
