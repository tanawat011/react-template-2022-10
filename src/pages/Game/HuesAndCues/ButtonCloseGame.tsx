import { useRecoilState } from 'recoil'

import { Button } from 'components/Button'
import { useSessionStorage } from 'hooks'
import { huesAndCuesRoomAtom, huesAndCuesRoomPlayersAtom } from 'recoils/huesAndCues'

import { deleteAllGlobalPlayers, deleteRoom } from './services'

export const ButtonCloseGame: React.FC = () => {
  const [, , deleteRoomId] = useSessionStorage('roomId')
  const [, , deleteMeId] = useSessionStorage('meId')

  const [room] = useRecoilState(huesAndCuesRoomAtom)
  const [players] = useRecoilState(huesAndCuesRoomPlayersAtom)

  const handleClickCloseGame = async () => {
    await deleteRoom(room.id)
    await deleteAllGlobalPlayers(players)

    deleteRoomId()
    deleteMeId()
  }

  return <Button onClick={handleClickCloseGame}>Close Game</Button>
}
