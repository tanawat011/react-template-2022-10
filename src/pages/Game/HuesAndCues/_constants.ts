export const SESSION = {
  ID: 'sessionId',
  ROOM_ID: 'roomId',
  TEMP_ID: 'tempSessionId',
}

export const COLLECTION = {
  HUES_AND_CUES: 'hues_and_cues',
  PLAYERS: 'players',
}

export const DOCUMENT = {
  ROOMS: 'rooms',
}
