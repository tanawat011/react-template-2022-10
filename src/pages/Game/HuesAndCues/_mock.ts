import type { RoomPlayer } from './_type'

const PLAYER_BASE: RoomPlayer = {
  cells: [],
  uid: '1',
  name: 'Player 1',
  score: 0,
  color: '',
  number: 1,
}

const result = 'F4'
const choice = ['F4', 'A9', 'P27', 'E15']

export const MOCK: {
  ONE_PLAYER: RoomPlayer[]
  TWO_PLAYER: RoomPlayer[]
  TWO_PLAYER_LAST_IS_HINTER: RoomPlayer[]
  THREE_PLAYER: RoomPlayer[]
  THREE_PLAYER_SECOND_IS_HINTER: RoomPlayer[]
  THREE_PLAYER_LAST_IS_HINTER: RoomPlayer[]
  THREE_PLAYER_LAST_IS_TURN: RoomPlayer[]
  HINTER_NO_CHOICE: RoomPlayer[]
  HINTER_NO_SELECT_CHOICE: RoomPlayer[]
  HINTER_SELECTED_CHOICE: RoomPlayer[]
  PLAYER_NO_COLOR: RoomPlayer[]
  PLAYER_HAS_COLOR: RoomPlayer[]
  PLAYER_HAS_COLOR_START_GAME: RoomPlayer[]
  PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT: RoomPlayer[]
} = {
  ONE_PLAYER: [
    {
      ...PLAYER_BASE,
      cells: ['D17'],
      number: 1,
      hinter: true,
      isMyTurn: true,
      color: 'bg-[#0758a5]',
    },
  ],
  TWO_PLAYER: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
  ],
  TWO_PLAYER_LAST_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      hinter: true,
    },
  ],
  THREE_PLAYER: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      result,
      choice,
      owner: true,
      score: 1,
    },
    {
      ...PLAYER_BASE,
      uid: '2',
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      uid: '3',
      number: 3,
    },
  ],
  THREE_PLAYER_LAST_IS_TURN: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      result,
      choice,
    },
    {
      ...PLAYER_BASE,
      number: 2,
    },
    {
      ...PLAYER_BASE,
      number: 3,
      isMyTurn: true,
    },
  ],
  THREE_PLAYER_SECOND_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  THREE_PLAYER_LAST_IS_HINTER: [
    {
      ...PLAYER_BASE,
      number: 1,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
      hinter: true,
    },
  ],
  HINTER_NO_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  HINTER_NO_SELECT_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      choice,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  HINTER_SELECTED_CHOICE: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      choice,
      result,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_NO_COLOR: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_HAS_COLOR: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
      color: 'bg-[#117822]',
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
  PLAYER_HAS_COLOR_START_GAME: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
      isStartGame: true,
      result: 'F4',
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
      cells: ['F4'],
    },
    {
      ...PLAYER_BASE,
      number: 3,
      cells: ['F5', 'F6'],
    },
  ],
  PLAYER_HAS_COLOR_START_GAME_SUBMIT_RESULT: [
    {
      ...PLAYER_BASE,
      number: 1,
      hinter: true,
      color: 'bg-[#0758a5]',
      owner: true,
      isStartGame: true,
      isSubmitResult: true,
      countRound: 1,
    },
    {
      ...PLAYER_BASE,
      number: 2,
      isMyTurn: true,
    },
    {
      ...PLAYER_BASE,
      number: 3,
    },
  ],
}
