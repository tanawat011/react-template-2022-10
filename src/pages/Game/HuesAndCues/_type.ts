export type Player = {
  uid: string
  name: string
}

export type RoomPlayer = {
  uid: string
  name: string
  score: number
  color: string // * color is a hex value
  cells: string[] // * cells that the player has selected
  owner: boolean // * if true, the player is the owner of the room
  number: number // * sequential number of player in room
  hinter?: boolean // * is the player the hinter
  choice?: string[] // * choice is an array of colorIds that the player can select
  result?: string // * result is the colorId of the color that the player has selected
  isSubmitResult?: boolean // * If true, the player has responded to the hint
  isMyTurn?: boolean // * If true, it is the player's turn
  isStartGame?: boolean // * If true, the game has started
  countTurn?: number // * The number of turns the player has taken
  countRound?: number // * The number of rounds that have passed
}
