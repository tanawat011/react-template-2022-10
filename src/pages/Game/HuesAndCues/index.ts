export * from './HuesAndCues'
export * from './SetDisplayName'
export * from './SetupRoom'
