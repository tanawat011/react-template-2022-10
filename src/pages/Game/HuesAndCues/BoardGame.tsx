import { useRecoilState } from 'recoil'
import tw from 'twin.macro'

import { TwCard } from 'components/Common'
import { getColTitles, getRowTitles, huesAndCues } from 'data/HuesAndCues'
import { huesAndCuesRoomAtom, huesAndCuesRoomPlayersAtom } from 'recoils/huesAndCues'

const BoardGameCard = tw(TwCard)`col-span-2 row-span-3 grid gap-2`
const BoardGameWrap = tw.div`m-auto grid gap-2`
const BoardGameMiddleSection = tw.div`grid grid-flow-col gap-2`
const BoardGameContent = tw.div`grid gap-2`
const CellWrap = tw.div`grid grid-flow-col gap-2`
const CellBG = tw.div`h-6 w-6 relative`
const Cell = tw(CellBG)`z-10 p-1 cursor-pointer hover:p-0.5 hover:border-2 hover:border-white`
const ColTitlesWrap = tw.div`grid grid-flow-col gap-2`
const ColTitle = tw.div`h-6 w-6 text-center text-white font-bold`
const RowTitlesWrap = tw.div`grid gap-2`
const RowTitle = tw.div`h-6 w-6 text-center text-white font-bold`
const SelectorToken = tw.div`w-4 h-4 rounded-full border border-white `
const HintResult = tw.div`absolute border-4 border-black rounded h-24 w-24 -top-9 -left-9 bg-transparent`

const ColTitles = () => (
  <ColTitlesWrap>
    {['', ...getColTitles(), ''].map((title, index) => (
      <ColTitle key={`col-title-${title}-${index}`}>{title}</ColTitle>
    ))}
  </ColTitlesWrap>
)

const RowTitles = () => (
  <RowTitlesWrap>
    {getRowTitles().map((title, index) => (
      <RowTitle key={`row-title-${title}-${index}`}>{title}</RowTitle>
    ))}
  </RowTitlesWrap>
)

export const BoardGame = () => {
  const [room] = useRecoilState(huesAndCuesRoomAtom)
  const [roomPlayers] = useRecoilState(huesAndCuesRoomPlayersAtom)

  return (
    <BoardGameCard>
      <BoardGameWrap>
        <ColTitles />

        <BoardGameMiddleSection>
          <RowTitles />

          <BoardGameContent>
            {huesAndCues.map((items, index) => {
              return (
                <CellWrap key={`hues-and-cues-${index}`}>
                  {items.map((item) => {
                    const itemId = item.id
                    const itemColor = item.color

                    const roomPlayerSelected = roomPlayers.find((rp) =>
                      rp.allSelected.includes(itemId),
                    )

                    const isRoomResult = room.hintSelected === itemId

                    return (
                      <CellBG key={`${itemColor}-${itemId}`}>
                        <Cell className={itemColor}>
                          {roomPlayerSelected && (
                            <SelectorToken className={roomPlayerSelected.color} />
                          )}
                        </Cell>

                        {isRoomResult && <HintResult />}
                      </CellBG>
                    )
                  })}
                </CellWrap>
              )
            })}
          </BoardGameContent>

          <RowTitles />
        </BoardGameMiddleSection>

        <ColTitles />
      </BoardGameWrap>
    </BoardGameCard>
  )
}
