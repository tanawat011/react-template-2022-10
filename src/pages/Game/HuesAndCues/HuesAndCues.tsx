import type { RoomPlayer } from './_type'
import type { Unsubscribe } from 'firebase/firestore'

import { useEffect, useState } from 'react'

import { useNavigate, useParams } from 'react-router-dom'
import { v4 as uuid } from 'uuid'

import { TwCol } from 'components/Common'
import { useSessionStorage } from 'hooks'

import { Board } from './Board'
import { ChooseColorHint } from './ChooseColorHint'
import { ChooseColorToken } from './ChooseColorToken'
import { NextHinter } from './NextHinter'
import { NextRound } from './NextRound'
import { PlayerColor } from './PlayerColor'
import { RemoveRoom } from './RemoveRoom'
import { ResetRoom } from './ResetRoom'
import { ScoreBoard } from './ScoreBoard'
import { StartGame } from './StartGame'
import { SubmitResult } from './SubmitResult'
import { SESSION } from './_constants'
import {
  getAllPlayersInTheRoom,
  getPlayerInTheRoom,
  setPlayerInTheRoom,
  subscribeRoom,
  toSetDisplayRoom,
} from './services'

export const HuesAndCues = () => {
  const urlParam = useParams<{ roomId: string }>()
  const navigate = useNavigate()

  const [sessionId, , removeSessionId] = useSessionStorage(SESSION.ID)
  const [roomId, setRoomId, removeRoomId] = useSessionStorage(SESSION.ROOM_ID)
  const [tempSessionId, setTempSessionId, removeTempSessionId] = useSessionStorage(SESSION.TEMP_ID)

  const [playersInRoom, setPlayersInRoom] = useState<RoomPlayer[]>([])
  const [currentPlayer, setCurrentPlayer] = useState<RoomPlayer>()

  let unsubscribe: Unsubscribe = () => {
    return
  }

  useEffect(() => {
    checkRoomAndFetchData()

    return () => {
      unsubscribe()
    }
  }, [])

  const checkRoomAndFetchData = async () => {
    const _roomId = urlParam.roomId || ''

    const _playersInRoom = await getAllPlayersInTheRoom(_roomId)

    if (!_playersInRoom.length) {
      removeSessionId()
      removeRoomId()
      removeTempSessionId()

      toSetDisplayRoom(navigate)

      return
    }

    if (!sessionId) {
      const _sessionId = tempSessionId || uuid()

      setRoomId(_roomId)
      setTempSessionId(_sessionId)

      toSetDisplayRoom(navigate)

      return
    }

    await fetchCurrentPlayer()

    fetchPlayersInRoom()
  }

  const fetchCurrentPlayer = async () => {
    const data = await getPlayerInTheRoom(roomId, sessionId)

    setCurrentPlayer(data)
  }

  const updateCurrentPlayer = async (_currentPlayer: RoomPlayer) => {
    setCurrentPlayer(_currentPlayer)

    await setPlayerInTheRoom(roomId, _currentPlayer)
  }

  const fetchPlayersInRoom = () => {
    const callback = (data: RoomPlayer[]) => {
      setPlayersInRoom(data)
      fetchCurrentPlayer()
    }

    unsubscribe = subscribeRoom(roomId, callback)
  }

  const isOwnerRoom = currentPlayer && currentPlayer.owner
  const isStartGame = playersInRoom.findIndex((p) => p.isStartGame) > -1
  const isHinter = currentPlayer && currentPlayer.hinter
  const isPlayerNoColor = currentPlayer && !currentPlayer.color
  const isPlayerHasColor = currentPlayer && currentPlayer.color

  return (
    <TwCol className='p-6'>
      {isOwnerRoom && (
        <>
          <RemoveRoom roomId={roomId} />

          <ResetRoom
            roomId={roomId}
            currentPlayer={currentPlayer}
            setCurrentPlayer={setCurrentPlayer}
          />

          <StartGame
            roomId={roomId}
            playersInRoom={playersInRoom}
            currentPlayer={currentPlayer}
            updateCurrentPlayer={updateCurrentPlayer}
          />
        </>
      )}

      {currentPlayer?.hinter && (
        <>
          <NextHinter
            roomId={roomId}
            playersInRoom={playersInRoom}
            currentPlayer={currentPlayer}
            updateCurrentHinter={updateCurrentPlayer}
          />

          <SubmitResult
            roomId={roomId}
            playersInRoom={playersInRoom}
            currentPlayer={currentPlayer}
            updateCurrentPlayer={updateCurrentPlayer}
          />

          <NextRound
            roomId={roomId}
            playersInRoom={playersInRoom}
            currentPlayer={currentPlayer}
            setCurrentPlayer={setCurrentPlayer}
          />
        </>
      )}

      <ScoreBoard playersInRoom={playersInRoom} />

      {isPlayerNoColor && (
        <ChooseColorToken
          currentPlayer={currentPlayer}
          playersInRoom={playersInRoom}
          updateCurrentPlayer={updateCurrentPlayer}
        />
      )}

      {isPlayerHasColor && (
        <>
          <PlayerColor
            currentPlayer={currentPlayer}
            playersInRoom={playersInRoom}
            updateCurrentPlayer={updateCurrentPlayer}
          />

          {isHinter && (
            <ChooseColorHint
              currentPlayer={currentPlayer}
              updateCurrentPlayer={updateCurrentPlayer}
            />
          )}

          {isStartGame && (
            <Board
              roomId={roomId}
              currentPlayer={currentPlayer}
              playersInRoom={playersInRoom}
              updateCurrentPlayer={updateCurrentPlayer}
            />
          )}
        </>
      )}
    </TwCol>
  )
}
