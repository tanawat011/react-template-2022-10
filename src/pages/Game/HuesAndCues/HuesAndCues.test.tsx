import { renderWithProviders } from 'helpers/test'

import { GameHuesAndCues } from './HuesAndCues'

describe('<GameHuesAndCues />', () => {
  test('match snapshot', () => {
    const { asFragment } = renderWithProviders(<GameHuesAndCues />)
    expect(asFragment()).toMatchSnapshot()
  })
})
