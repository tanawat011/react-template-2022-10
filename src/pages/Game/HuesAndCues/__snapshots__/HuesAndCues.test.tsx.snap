// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameHuesAndCues /> match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c2 {
  margin: 0.25rem;
  width: 1.75rem;
  height: 1.75rem;
  --tw-text-opacity: 1;
  color: rgb(255 255 255 / var(--tw-text-opacity));
  font-weight: 700;
  text-align: center;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
}

.c3 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 0.75rem;
  line-height: 1rem;
}

.c4 {
  width: 2.25rem;
  height: 3.5rem;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c5 {
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

<div
    class="c0 p-6"
  >
    <div
      class="mb-2 p-3 shadow-around"
    >
      <div
        class="pl-2 mb-1 text-white font-bold select-none"
      >
        Score Board
      </div>
      <div
        class="c0 items-start"
      >
        <div
          class="c1"
        >
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#2e2c2d]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#323031]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#353132]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#393536]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            >
              5
            </div>
            <div
              class="c2 c3 c4 bg-[#3a3539]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#403c3d]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#423e3f]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#474342]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#484745]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            >
              10
            </div>
            <div
              class="c2 c3 c4 bg-[#4d4948]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#4e4d4b]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#545051]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#565052]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#575354]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            >
              15
            </div>
            <div
              class="c2 c3 c4 bg-[#595556]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#615b5b]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#605a5a]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#65605d]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#686262]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            >
              20
            </div>
            <div
              class="c2 c3 c4 bg-[#676364]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#6a6667]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#696566]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#6d676b]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            />
            <div
              class="c2 c3 c4 bg-[#6c676b]"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="text-center text-white font-semibold h-6"
            >
              25
            </div>
            <div
              class="c2 c3 c4 bg-[#6d686c]"
            />
          </div>
        </div>
        <div
          class="c1 c5"
        >
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#757074]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#787475]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#7e787a]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#817b7b]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#86817e]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            >
              30
            </div>
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#898481]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#8e8684]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#918986]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#968e8b]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#97928d]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            >
              35
            </div>
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#97928e]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#9d9592]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#9e9693]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#a19996]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#a09b97]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            >
              40
            </div>
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#a69e9b]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#a5a09c]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#a7a19e]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#a6a19d]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#a5a29d]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            >
              45
            </div>
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#a7a29e]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#a6a39e]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#aca8a5]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#a6a29f]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            />
          </div>
          <div
            class="c0"
          >
            <div
              class="c2 c3 c4 bg-[#a7a3a0]"
            />
            <div
              class="text-center text-white font-semibold h-6"
            >
              50
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
